File: app/src/main/java/com/udacity/astroapp/utils/Converters.java
package com.udacity.astroapp.utils;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.udacity.astroapp.models.Camera;
import com.udacity.astroapp.models.MarsPhotoObject;
import com.udacity.astroapp.models.Rover;

import java.lang.reflect.Type;

/**
 * Converters for saving the custom objects to the Room database
 */
public class Converters {

    @TypeConverter
    public static Camera cameraFromString(String cameraValue) {
        Type cameraType = new TypeToken<Camera>() {
        }.getType();
        return new Gson().fromJson(cameraValue, cameraType);
    }

    @TypeConverter
    public static String cameraToString(Camera camera) {
        return new Gson().toJson(camera);
    }

    @TypeConverter
    public static Rover roverFromString(String roverValue) {
        Type roverType = new TypeToken<Rover>() {
        }.getType();
        return new Gson().fromJson(roverValue, roverType);
    }

    @TypeConverter
    public static String roverToString(Rover rover) {
        return new Gson().toJson(rover);
    }
}

File: app/src/main/kotlin/com/udacity/astroapp/utils/Converters.kt
package com.udacity.astroapp.utils

import androidx.room.TypeConverter
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.udacity.astroapp.models.Camera
import com.udacity.astroapp.models.Rover

/** Converters for saving the custom objects to the Room database */
object Converters {

    @TypeConverter
    @JvmStatic
    fun cameraFromString(cameraValue: String?): Camera? {
        if (cameraValue.isNullOrEmpty()) return null
        val cameraType = object : TypeToken<Camera>() {}.type
        return Gson().fromJson(cameraValue, cameraType)
    }

    @TypeConverter
    @JvmStatic
    fun cameraToString(camera: Camera?): String? {
        return camera?.let { Gson().toJson(it) }
    }

    @TypeConverter
    @JvmStatic
    fun roverFromString(roverValue: String?): Rover? {
        if (roverValue.isNullOrEmpty()) return null
        val roverType = object : TypeToken<Rover>() {}.type
        return Gson().fromJson(roverValue, roverType)
    }

    @TypeConverter
    @JvmStatic
    fun roverToString(rover: Rover?): String? {
        return rover?.let { Gson().toJson(it) }
    }
}
